### PR TITLE
MultiCombobox: Add basic controlled functionality

### PR DESCRIPTION
**Pull request by tskarhed**
**What is this feature?**

Adds basic functionality for `value` and `onChange` props

**Why do we need this feature?**

Otherwise consumers can't use it.

**Who is this feature for?**

Everyone

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import React from 'react';
+
+import { MultiCombobox, MultiComboboxProps } from './MultiCombobox';
+
+describe('MultiCombobox', () => {
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:7]: Nice tests 👍

**Review by joshhunt — COMMENTED** [review]
just my quick once over. Will check it out in detail later 👍

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx [hunk 2]
```diff
@@ -11,22 +11,52 @@ import { ValuePill } from './ValuePill';
...
-  const { options, placeholder } = props;
+  const { options, placeholder, onChange, value } = props;
+  const isAsync = typeof options === 'function';
 
-  const multiStyles = useStyles2(getMultiComboboxStyles);
+  const selectedItems = useMemo(() => {
+    if (!value || isAsync) {
+      //TODO handle async
+      return [];
+    }
```
[the remark this answers is not in the document]
Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx:28]: This is a side effect from extending the original Combobox interface, although it leaves a lot of unimplemented functionality.

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx [hunk 3]
```diff
@@ -11,22 +11,52 @@ import { ValuePill } from './ValuePill';
...
+
+    return value;
+  }, [value, options, isAsync]);
+
+  const multiStyles = useStyles2(getMultiComboboxStyles);
 
   const [items, _baseSetItems] = useState(isAsync ? [] : options);
   const [isOpen, setIsOpen] = useState(false);
-  const [selectedItems, setSelectedItems] = useState<Array<ComboboxOption<T>>>([]);
+  //const [selectedItems, setSelectedItems] = useState<Array<ComboboxOption<T>>>(initialSelectedItems);
```
Comment (eledobleefe) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx:54]: Was this commented while adding the `selectedItems` on line 24? Do you still need this `selectedItems`?

Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx:54]: No, it was a little oopsie to leave the comment there 😅

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    const options = [
+      { label: 'A', value: 'a' },
+      { label: 'B', value: 'b' },
+      { label: 'C', value: 'c' },
+    ];
+    render(<MultiCombobox options={options} value={[]} onChange={jest.fn()} placeholder="Select" />);
+    expect(screen.getByPlaceholderText('Select')).toBeInTheDocument();
+  });
+
+  it.each([
```
Comment (eledobleefe) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:42]: TIL 🙂

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    //Deselect
+    //userEvent.click(input);
+    await userEvent.click(await screen.findByRole('option', { name: 'A' }));
+
+    expect(onChange).toHaveBeenNthCalledWith(1, [first]);
+    expect(onChange).toHaveBeenNthCalledWith(2, [first, third]);
+    expect(onChange).toHaveBeenNthCalledWith(3, [third]);
+  });
+
+  it('should be able to render a valie that is not in the options', async () => {
```
Comment (eledobleefe) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:85]:
```suggestion
  it('should be able to render a value that is not in the options', async () => {
```

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx [hunk 6]
```diff
@@ -55,26 +92,37 @@ export const MultiCombobox = <T extends string | number>(props: MultiComboboxPro
...
     getItemProps,
-    //selectedItem,
   } = useCombobox({
     isOpen,
     items,
     itemToString,
     inputValue,
-    //defaultHighlightedIndex: 0,
     selectedItem: null,
+    stateReducer: (state, actionAndChanges) => {
```
Comment (eledobleefe) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx:101]: The 'state' seems to be not used 🤔

**Review by eledobleefe — APPROVED** [review]
As far as I know, looks good 👌

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+describe('MultiCombobox', () => {
+  it('should render with options', async () => {
+    const options = [
+      { label: 'A', value: 'a' },
+      { label: 'B', value: 'b' },
+      { label: 'C', value: 'c' },
+    ];
+    render(<MultiCombobox options={options} value={[]} onChange={jest.fn()} />);
+    const input = screen.getByRole('combobox');
+    userEvent.click(input);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:16]: Nit: It's better to [set up the userEvent](https://testing-library.com/docs/user-event/setup/), as calling it directly might not work in the future versions.

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    expect(await screen.findByText('C')).toBeInTheDocument();
+  });
+
+  it('should render with value', () => {
+    const options = [
+      { label: 'A', value: 'a' },
+      { label: 'B', value: 'b' },
+      { label: 'C', value: 'c' },
+    ];
+    const { getByText } = render(<MultiCombobox options={options} value={['a']} onChange={jest.fn()} />);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:28]: I think the current recommendation is to call these methods from `screen`, e.g., `screen.getByText`.

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    await userEvent.click(input);
+    await userEvent.click(await screen.findByRole('option', { name: 'A' }));
+
+    //Second option
+    //await userEvent.click(input);
+    await userEvent.click(await screen.findByRole('option', { name: 'C' }));
+
+    //Deselect
+    //userEvent.click(input);
+    await userEvent.click(await screen.findByRole('option', { name: 'A' }));
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:78]:
```suggestion
    await userEvent.click(screen.getByRole('option', { name: 'C' }));

    //Deselect
    //userEvent.click(input);
    await userEvent.click(screen.getByRole('option', { name: 'A' }));
```
It shouldn't be necessary to use `findBy` in subsequent queries (only the first one), as by that time all async actions should be already resolved.

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx [hunk 10]
```diff
@@ -11,22 +11,58 @@ import { ValuePill } from './ValuePill';
...
+      for (const [index, val] of value.entries()) {
+        if (resultingItems[index] === undefined) {
+          resultingItems[index] = { value: val };
+        }
+      }
+      return resultingItems.filter((item) => item !== undefined); // TODO: Not actually needed, but TS complains
+    }
+
+    return value;
+  }, [value, options, isAsync]);
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Combobox/MultiCombobox.tsx:55]: Would moving this logic into a separate function make sense?

File: packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx (hunk 8 above)
Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/MultiCombobox.test.tsx:28]: Something I need to teach copilot ;)

**Pull request merged**